src/utilities/device_utils.c

```diff
@@ -800,8 +800,11 @@ hypreDevice_CsrRowIndicesToPtrs_v2( HYPRE_Int  nrows,
                                     HYPRE_Int *d_row_ptr )
 {
 #if defined(HYPRE_USING_SYCL)
-   if (nnz <= 0 || nrows <= 0)
+   /* WM: if nnz <= 0, then dpl::lower_bound is a no-op, which means we still need to zero out the row pointer */
+   /* WM: question - does thrust lower_bound do this automatically? Otherwise, I'd expect this to break for cuda also. */
```

Comment:
@liruipeng , do you have an answer to this? I ran into an issue with the sycl build where hypreDevice_CsrRowIndicesToPtrs_v2() was called with nnz = 0 and nrows nonzero. In this case, the dpl lower_bound() call was a no-op, so d_row_ptr was left unchanged (with incorrect nonzero entries), leading to errors later on. What is the behavior of thrust's lower_bound in this case? If it is also a no-op, I would expect to need this check for nnz <= 0 for both cuda and sycl.

Reply:
OK, I tested this on lassen and verified that, for the cuda build, this function behaves correctly even when nnz = 0 and nrows is nonzero. That is, when nnz = 0, lower_bound sets the row pointer to 0 everywhere. This is NOT true for oneDPL's lower_bound, so for the sycl backend we need to explicitly take care of this case.

Reply:
@waynemitchell Yes, you are right. I think `thrust::lower_bound` is consistent with C++ standard behavior (see, https://cplusplus.com/reference/algorithm/lower_bound/, look at the "The behavior of this function template is equivalent to:" part). That is, when `first == last`, it should return `first`, which is `0` here. But treating the special case (`nnz == 0`) separately as you did looks fine to me (even more clear).